runtime/compiler/codegen/J9CodeGenerator.cpp

```diff
@@ -1493,6 +1500,201 @@ J9::CodeGenerator::lowerTreeIfNeeded(
+      //   |   <array-reference>            |
+      //   | ificmpeq  -->------------------*---------+
+      //   |   iand                         |         |
+      //   |     iloadi <isClassFlags>      |         |
+      //   |       aloadi <componentClass>  |         |
+      //   |         aloadi <vft-symbol>    |         |
+      //   |           <array-reference>    |         |
+      //   |     iconst J9ClassIsValueType  |         |
+      //   |   iconst 0                     |         |
+      //   | BBEnd                          |         |
+      //   +--------------------------------+         |
+      //   | BBStart (Extension)            |         |
+      //   | NULLCHK                        |         |
+      //   |   Passthrough                  |         |
+      //   |     <value-reference>          |         |
+      //   | BBEnd                          |         |
+      //   +--------------------------------+         |
+      //                   |                          |
+      //                   +--------------------------+
+      //                   |
+      //                   v
+      //   +--------------------------------+
+      //   | BBStart                        |
+      //   | ArrayStoreCHK                  |
+      //   |   awrtbari                     |
+      //   |     aladd                      |
+      //   |       aload <array>            |
+      //   |       index-offset-calculation |
+      //   |     aload <value>              |
+      //   +--------------------------------+
+      //
+      TR::SymbolReference *vftSymRef = self()->comp()->getSymRefTab()->findOrCreateVftSymbolRef();
+      TR::SymbolReference *arrayCompSymRef = self()->comp()->getSymRefTab()->findOrCreateArrayComponentTypeSymbolRef();
+      TR::SymbolReference *classFlagsSymRef = self()->comp()->getSymRefTab()->findOrCreateClassFlagsSymbolRef();
+
+      TR::Node *vft = TR::Node::createWithSymRef(node, TR::aloadi, 1, anchoredArrayTT->getNode()->getFirstChild(), vftSymRef);
+      TR::Node *arrayCompClass = TR::Node::createWithSymRef(node, TR::aloadi, 1, vft, arrayCompSymRef);
+      TR::Node *loadClassFlags = TR::Node::createWithSymRef(node, TR::iloadi, 1, arrayCompClass, classFlagsSymRef);
+      TR::Node *isValueTypeNode = TR::Node::create(node, TR::iand, 2, loadClassFlags, TR::Node::iconst(node, J9ClassIsValueType));
+
+      TR::Node *ifNode = TR::Node::createif(TR::ificmpeq, isValueTypeNode, TR::Node::iconst(node, 0));
+      ifNode->copyByteCodeInfo(node);
+
+
+      TR::Node *passThru  = TR::Node::create(node, TR::PassThrough, 1, sourceChild);
+      TR::ResolvedMethodSymbol *currentMethod = self()->comp()->getMethodSymbol();
+      TR::Node *nullCheck = TR::Node::createWithSymRef(node, TR::NULLCHK, 1, passThru,
+                               symRefTab()->findOrCreateNullCheckSymbolRef(currentMethod));
+
+      TR::Block *nullCheckBlock = prevBlock->splitPostGRA(tt, cfg);
```

Comment:
@hzongaro  Looking at the control flow described in the comment, I think you are using wrong API here. Block that contains `ArrayStoreCHK` tree is the merge block, so when creating `arrayStoreCheckBlock` I would use `splitPostGRA`. This way we have done uncommoning of the nodes that are live in the merge block. While for creating block containing `NULLCHK` which seems like the extension of previous block, I would use regular `split` API with `doCommoning` boolean set to false. 
 For reference on copying `GlRegDeps` you can take a look at  https://github.com/eclipse/openj9/blob/a7bf6c09925da0a4e352114e63b3c2cb3f59b355/runtime/compiler/optimizer/JProfilingValue.cpp#L764-L781.

Reply:
Thanks for your review and advice, Rahil @r30shah - revised in commit https://github.com/eclipse/openj9/commit/33c4ed812005bc98d94f4fe9282a75b0aaf0a6f6

Reply:
Updated commit to https://github.com/eclipse/openj9/commit/58ab83bcf1cf130105931a3c27f9fa299cc3acd4 - no further changes there with respect to this comment.